### PR TITLE
feat(transformer): implement typescript namespace

### PR DESCRIPTION
**Pull request by Boshen**

### CODE REVIEW

**Comment by Dunqing** [comment]
Most of the remaining tests can be solved after we implement the `generateUid` function. Due to the inconsistency between our `enum` implementation and babel's, the test case about `enum` also fails to pass the.

``````
namespace/nested-destructuring/input.ts

**Comment by Dunqing** [comment]
The PR is done, Please convert to open and review this! Thanks! @Boshen

File: crates/oxc_transformer/src/typescript/namespace.rs [hunk 1]
```diff
@@ -11,99 +11,105 @@ use oxc_syntax::operator::{AssignmentOperator, LogicalOperator};
 // 1. register scope for the newly created function: <https://github.com/babel/babel/blob/08b0472069cd207f043dd40a4d157addfdd36011/packages/babel-plugin-transform-typescript/src/namespace.ts#L38>
 impl<'a> TypeScript<'a> {
     // `namespace Foo { }` -> `let Foo; (function (_Foo) { })(Foo || (Foo = {}));`
-    pub(super) fn transform_statements_for_namespace(&self, stmts: &mut Vec<'a, Statement<'a>>) {
+    pub(super) fn transform_program_for_namespace(&self, program: &mut Program<'a>) {
+        if !has_namespace(program.body.as_slice()) {
```
Comment (Boshen) [inline crates/oxc_transformer/src/typescript/namespace.rs:15]: I was confused about this, you should comment on namespace can only be at top level.

**Comment by Boshen** [comment]
Feel free to merge after another round of reviews by yourself.

**Comment by Dunqing** [comment]
* **#3025** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/3025?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Boshen and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by Dunqing — APPROVED** [review]

**Pull request merged**